Updated kaslr-fix.md (#24)
macos86/macos86.github.io: extras/kaslr-fix.md

* **ProvideCustomSlide**: YES
  * Questo assicura che il kernel scelga solo regioni buone ed eviti quelle che potrebbero causare errori di avvio. È ancora casuale ma omette quelle regioni difettose nella sua randomizzazione
* **RebuildAppleMemoryMap**: YES
  * Genera una mappa di memoria compatibile con macOS in modo che si adatterà a ciò che il kernel si aspetta. Su alcuni firmware OEM di laptop potresti non essere in grado di avviare con questa opzione attiva. Nel caso disattivala.

> ## To fix ^^

Author: Some OEM laptops firmwares may mesbehave with `RebuildAppleMemoryMap` enabled (such as non bootable system usually). If the computer doesn't start at all you need to disable this option as well as `SyncRuntimePermissions`.

So theorically you can rewrite the phrase with:

`Genera un memory layout compatibile con macOS. N.B. su alcuni firmware OEM di laptop potrebbero esserci dei problemi con questo quirk. Nel caso disattivarlo così come SyncRuntimePermissions`

Let me know your thoughts

> ## To fix ^^

Reviewer: OK it can be done!

## Preparazione del BIOS

Il motivo per cui abbiamo bisogno di ripristinare la mappa di memoria è che vogliamo che sia più deterministico,ossia che ci saranno meno variazioni su ogni avvio, quindi abbiamo meno casi limite (le mappe di memoria non calzano sempre a pennello). Per iniziare:

* Aggiorna il BIOS all'ultima versione STABILE (i primi BIOS spediti sono noti per avere problemi con il memory layout, specialmente con Z390)
* Resettare il CMOS
* Abilitare alcune impostazioni del BIOS necessarie:

>   * `Above4GDecoding`: consente ai dispositivi PCI di utilizzare regioni di memoria superiori a 4 GB, il che significa che macOS avrà più spazio per adattarsi; può essere problematico su alcuni chipset come X99 e X299 quindi consigliamo di testare con e senza questa opzione attivata.

Reviewer: ```suggestion
  * `Above4GDecoding`: consente ai dispositivi PCI di utilizzare regioni di memoria superiori a 4 GB, il che significa che macOS avrà più spazio per adattarsi; può essere problematico su alcuni chipset come X99 e X299 quindi consigliamo di attivarla e in caso di problemi disattivarla.
```

>     * Nota: sui BIOS del 2020 e successivi è stato introdotto il supporto al `Resizable BAR`, opzione che si sblocca con l'abilitazione di `Above4GDecoding`. Assicurarsi che il supporto al `Resizable BAR` sia disabilitato se l'opzione si presenta.

Reviewer: ```suggestion
    * Nota: Dal 2020 nei BIOS è stato introdotto il supporto al `Resizable BAR`, opzione che si sblocca con l'abilitazione di `Above4GDecoding`. Assicurarsi che il supporto al `Resizable BAR` sia disabilitato se l'opzione si presenta.
```

  * `Opzioni di avvio -> Modalità Windows 8.1/10`: ci assicurerà che non verranno caricate delle opzioni per dispositivi legacy. Fun fact: `Other OS` è progettato solo per l'avvio di versioni precedenti di Windows e non per altri sistemi operativi.
* Disabilitare tutti i dispositivi non necessari nel BIOS (questo significa che ci sarà meno variazione nel memory layout ad ogni avvio, quindi meno possibilità di errori in fase di avvio). Impostazioni comuni:
  * `CSM`: per il supporto legacy, aggiunge un mucchio di spazzatura che non vogliamo. Anche questo può causare problemi in fase di avvio.
  * `Intel SGX`: Software Guard Extensions, occupa molto spazio e non ha alcuna utilità in macOS.
  * `Parallel Port`: macOS non può nemmeno sfruttare le porte parallele.

>   * `Serial Port`: Sarei curioso di sapere quanti di voi stiano debuggando il kernel...

Reviewer: ```suggestion
  * `Serial Port`: Sarei curioso di sapere quanti di voi le stiano usando per fare debug del kernel...
```

  * `iGPU`: Non è l'ideale, ma alcuni sistemi hanno mappe talmente grandi che l'iGPU non può adattarsi.

>   * `Thunderbolt`: Molti hack non hanno il supporto Thunderbolt funzionante; le schede che non hanno dispositivi Thunderbolt ma hanno questa opzione sprecano solo più spazio.

Reviewer: ```suggestion
  * `Thunderbolt`: Molti hack non supportano Thunderbolt; le schede che non hanno dispositivi Thunderbolt ma che hanno questa opzione sprecano solo più spazio.
```

  * `Illuminazione a LED`: Scusa amico, è ora di andare.
  * `USB legacy`: più schifezze legacy.

## Avvio di prova

Con le nostre impostazioni EFI, config.plist e BIOS modificate, è ora di provare la nostra nuova configurazione. Se hai ancora problemi, ci sarà da indagare ancora più a fondo per calcolare il valore di slide.

## Calcolare il valore dello slide

> Quello che dovrai fare è aprire una shell EFI (OpenShell, UEFIShell) nel tuo boot manager preferito ed eseguire `memmap`. L'output ti darà un elenco di tutte le pagine di memoria e delle loro relative dimensioni. È qui che inizia il divertimento.

Reviewer: ```suggestion
Quello che dovrai fare è aprire una shell EFI (OpenShell, UEFIShell) nel tuo boot manager preferito ed eseguire `memmap`. L'output ti darà un elenco di tutte le pagine di memoria e delle loro relative dimensioni. E qui inizia il divertimento:
```

| BS_Data | `000000006B526000` | `000000006B625FFF` | `0000000000000100` | `000000000000000F` |
| Available | `000000006B626000` | `000000006B634FFF` | `000000000000000F` | `000000000000000F` |

> Probabilmente ti starai chiedendo `come diavolo facciamo a convertirlo in un valore di Slide?`. Beh è abbastanza semplice. Quello che ci interessa è il valore più grande disponibile nella colonna `Start`. In questo esempio vediamo che `000000006B626000` è il nostro valore più grande (tieni presente che questi valori sono rappresentati in base esadecimale, quindi se ci sono più valori vicini tra loro potresti dover convertirli in base decimale). Per calcolare il valore dello slide basterà ricorrere a questa semplice formula matematica (la calcolatrice incorporata di macOS ha una funzione di programmazione premendo ⌘ + 3):

Reviewer: ```suggestion
Probabilmente ti starai chiedendo `come diavolo facciamo a convertirlo in un valore di Slide?`. Beh è abbastanza semplice. Quello che ci interessa è il valore più grande disponibile nella colonna `Start`. In questo esempio vediamo che `000000006B626000` è il nostro valore più grande (tieni presente che questi valori sono rappresentati in base esadecimale, quindi se ci sono più valori vicini tra loro potresti dover convertirli in base decimale). Per calcolare il valore dello slide basterà ricorrere a questa semplice formula matematica (la calcolatrice incorporata di macOS ha la sezione programmatore: abilitala premendo ⌘ + 3):
```

## Utilizzo di DevirtualiseMmio

> DevirtualiseMmio è un quirk piuttosto interessante, in quanto aggira un enorme ostacolo con molte schede madri con chipset Z390 e praticamente tutte le schede madri HDET con chipset come X99 e X299. Il modo in cui lo fa è che prende le regioni MMIO e rimuove gli attributi di runtime consentendo loro di essere utilizzati comodamente come spazio per il kernel, e accoppiarlo con il quirk di `ProvideCustomSlide` significa che possiamo mantenere le funzioni di sicurezza dello slide mentre otteniamo anche una macchina avviabile.

Reviewer: ```suggestion
DevirtualiseMmio è un quirk piuttosto interessante, in quanto aggira un enorme ostacolo con molte schede madri con chipset Z390 e praticamente tutte le schede madri HDET con chipset come X99 e X299. Il modo in cui lo fa è che prende le regioni MMIO e rimuove gli attributi di runtime consentendo loro di essere utilizzati comodamente come spazio per il kernel, e accoppiarlo con il quirk `ProvideCustomSlide` significa che possiamo mantenere le funzioni di sicurezza dello slide ottenendo anche una macchina avviabile.
```
